File: Mechanica Engine/src/matrices/ViewMatrix.java
package matrices;

import org.lwjgl.util.vector.Matrix4f;
import org.lwjgl.util.vector.Vector3f;

/**
 * Created by domin on 23 Mar 2017.
 */

public class ViewMatrix {
    private Matrix4f matrix;
    private Matrix4f matrixWithProjection;
    private Vector3f translation;
    private Vector3f xAxis = new Vector3f(1,0,0);
    private Vector3f yAxis = new Vector3f(0,1,0);
    private Vector3f zAxis = new Vector3f(0,0,1);
    private float positiionX = 0, positionY = 0, positionZ = 0;
    private float rx = 0, ry = 0, rz = 0;

    private boolean scheduleCreation = true;
    private boolean scheduleProjectionAndView = true;

    public ViewMatrix(){
        translation = new Vector3f();
        matrix = new Matrix4f();
        matrix.setIdentity();
        matrixWithProjection = new Matrix4f();
        matrixWithProjection.setIdentity();
        scheduleCreation = true;
    }

    public void setTranslate(float x, float y, float z){
        translation.set((float) - x, (float) - y, (float) - z);
        scheduleCreation = true;
    }

    public void setRotate(float x, float y, float z){
        rx = x;
        ry = y;
        rz = z;
        scheduleCreation = true;
    }

    public Matrix4f create(){
        if (scheduleCreation) {
            matrix.setIdentity();
            Matrix4f.translate(translation, matrix, matrix);

            Matrix4f.rotate((float) Math.toRadians(rx), xAxis, matrix, matrix);
            Matrix4f.rotate((float) Math.toRadians(ry), yAxis, matrix, matrix);
            Matrix4f.rotate((float) Math.toRadians(rz), zAxis, matrix, matrix);

            scheduleProjectionAndView = true;
        }
        scheduleCreation = false;
        return matrix;
    }

    public Matrix4f createWithProjection(ProjectionMatrix projectionMatrix){
        if (scheduleCreation || projectionMatrix.scheduleCreation || scheduleProjectionAndView) {
            matrix.setIdentity();
            Matrix4f.translate(translation, matrix, matrix);

            Matrix4f.rotate((float) Math.toRadians(rx), xAxis, matrix, matrix);
            Matrix4f.rotate((float) Math.toRadians(ry), yAxis, matrix, matrix);
            Matrix4f.rotate((float) Math.toRadians(rz), zAxis, matrix, matrix);

            Matrix4f.mul( projectionMatrix.create(), matrix, matrixWithProjection);
        }
        scheduleProjectionAndView = false;
        scheduleCreation = false;
        return matrixWithProjection;
    }
}

File: lwjgl3-test/src/matrices/ViewMatrix.kt
package matrices

import org.joml.Matrix4f
import org.joml.Vector3f

/**
 * Created by domin on 23 Mar 2017.
 */

class ViewMatrix {
    private val matrix = Matrix4f()
    private var matrixWithProjection = Matrix4f()
    private val translation: Vector3f = Vector3f()
    private val xAxis = Vector3f(1f, 0f, 0f)
    private val yAxis = Vector3f(0f, 1f, 0f)
    private val zAxis = Vector3f(0f, 0f, 1f)
    private var rx = 0f
    private var ry = 0f
    private var rz = 0f

    private var scheduleCreation = true
    private var scheduleProjectionAndView = true

    init {
        matrix.identity()
        matrixWithProjection.identity()
        scheduleCreation = true
    }

    fun setTranslate(x: Double, y: Double, z: Double) {
        translation.set((-x).toFloat(), (-y).toFloat(), (-z).toFloat())
        scheduleCreation = true
    }

    fun setRotate(x: Double, y: Double, z: Double) {
        rx = x.toFloat()
        ry = y.toFloat()
        rz = z.toFloat()
        scheduleCreation = true
    }

    fun create(): Matrix4f {
        if (scheduleCreation) {
            matrix.identity()
            matrix.translate(translation)

            matrix.rotate(Math.toRadians(rx.toDouble()).toFloat(), xAxis)
            matrix.rotate(Math.toRadians(ry.toDouble()).toFloat(), yAxis)
            matrix.rotate(Math.toRadians(rz.toDouble()).toFloat(), zAxis)

            scheduleProjectionAndView = true
        }
        scheduleCreation = false
        return matrix
    }

    fun createWithProjection(projectionMatrix: ProjectionMatrix): Matrix4f {
        if (scheduleCreation || projectionMatrix.scheduleCreation || scheduleProjectionAndView) {
            matrix.identity()
            matrix.translate(translation)

            matrix.rotate(Math.toRadians(rx.toDouble()).toFloat(), xAxis)
            matrix.rotate(Math.toRadians(ry.toDouble()).toFloat(), yAxis)
            matrix.rotate(Math.toRadians(rz.toDouble()).toFloat(), zAxis)

            matrixWithProjection = projectionMatrix.create().mul(matrix)
        }
        scheduleProjectionAndView = false
        scheduleCreation = false
        return matrixWithProjection
    }
}
